### PR TITLE
Add artifact registry locations support

### PR DESCRIPTION
**Pull request by attwad**
Add data source for artifact registry locations.

Fixes https://github.com/hashicorp/terraform-provider-google/issues/18785

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-datasource
`google_artifact_registry_locations`
```

### CODE REVIEW

**Comment by attwad** [comment]
Hi @NickElliot, this is my first contribution to terraform / magic-modules. I'm following https://googlecloudplatform.github.io/magic-modules/get-started/contribution-process/#contribute-code but the 3rd step (generate the providers) that I run with `make provider VERSION=beta OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google-beta" PRODUCT=artifactregistry` writes the changes in a different repo, I'm not sure how I am supposed to make this PR and that command work together, please advise, thank you!

**Review by NickElliot — COMMENTED** [review]
Hi, the idea is that when you run that command, it generates the changes into a TPG repo (via the same process that this repo leads to https://github.com/hashicorp/terraform-provider-google-beta downstream.

what you can then do is follow the following instructions to use your local version of TPG to run terraform https://googlecloudplatform.github.io/magic-modules/develop/test/run-tests/#optional-test-manually

Please let me know if you have any other questions!

**Comment by attwad** [comment]
Thanks for the explanation! I see some tests are failing (they passed locally even after I ran the provider make command which is weird) so I'll investigate.

**Review by NickElliot — CHANGES_REQUESTED** [review]
Problem is its failing lint :)

> gofmt needs running on the following files:
./google-beta/services/artifactregistry/data_source_artifact_registry_locations.go
./google-beta/services/artifactregistry/data_source_artifact_registry_locations_test.go
You can use the command: `make fmt` to reformat code.

**Comment by attwad** [comment]
Yup, I'm on vacation until August I'll fix that when I'm back, thanks!

**Review by NickElliot — COMMENTED** [review]
pending update

**Comment by attwad** [comment]
> Problem is its failing lint :)
> 
> > gofmt needs running on the following files:
> > ./google-beta/services/artifactregistry/data_source_artifact_registry_locations.go
> > ./google-beta/services/artifactregistry/data_source_artifact_registry_locations_test.go
> > You can use the command: `make fmt` to reformat code.

Formatted the code correctly now

File: mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_locations_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-provider-google/google/acctest"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
```
Comment (NickElliot) [inline mmv1/third_party/terraform/services/artifactregistry/data_source_artifact_registry_locations_test.go:12]:
```suggestion
	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
```

**Review by NickElliot — CHANGES_REQUESTED** [review]
try making this change thanks!

**Comment by attwad** [comment]
> try making this change thanks!

done but fmt reimported terraform, I've copied the imports from other tests as they looked different indeed, should be ok now

**Review by NickElliot — CHANGES_REQUESTED** [review]
same error is still present
> google-beta/services/artifactregistry/data_source_artifact_registry_locations_test.go:19:21: cannot use resource.TestCase{…} (value of type "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".TestCase) as "github.com/hashicorp/terraform-plugin-testing/helper/resource".TestCase value in argument to acctest.VcrTest

For this kind of issue it helps a lot if you can debug and run the test locally, as it costly in both our CI project, review time, and dramatically slows speed of PR merges to wait for the reviewer to manually be rerunning these tests, when basic build or syntax errors can often be caught really quickly locally. Thanks!

**Comment by attwad** [comment]
> same error is still present
> 
> > google-beta/services/artifactregistry/data_source_artifact_registry_locations_test.go:19:21: cannot use resource.TestCase{…} (value of type "github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource".TestCase) as "github.com/hashicorp/terraform-plugin-testing/helper/resource".TestCase value in argument to acctest.VcrTest
> 
> For this kind of issue it helps a lot if you can debug and run the test locally, as it costly in both our CI project, review time, and dramatically slows speed of PR merges to wait for the reviewer to manually be rerunning these tests, when basic build or syntax errors can often be caught really quickly locally. Thanks!

Well yes I've been running them, but the thing is they pass locally... :(

```
make provider VERSION=beta OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google-beta" PRODUCT=artifactregistry
cd "$GOPATH/src/github.com/hashicorp/terraform-provider-google-beta"
~/go/src/github.com/hashicorp/terraform-provider-google-beta$ go test google-beta/services/artifactregistry/data_source_artifact_registry_locations_test.go 
ok      command-line-arguments  0.126s
```

I'll keep looking, thanks

**Comment by attwad** [comment]
Even the tests requiring a gcloud project work:

GA:
```
~/go/src/github.com/hashicorp/terraform-provider-google$ make testacc TEST=./google/services/artifactregistry TESTARGS='-run=TestAccDataSourceGoogleArtifactRegistryLocations_basic'
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google/services/artifactregistry -v -run=TestAccDataSourceGoogleArtifactRegistryLocations_basic -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceGoogleArtifactRegistryLocations_basic
=== PAUSE TestAccDataSourceGoogleArtifactRegistryLocations_basic
=== CONT  TestAccDataSourceGoogleArtifactRegistryLocations_basic
--- PASS: TestAccDataSourceGoogleArtifactRegistryLocations_basic (6.15s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google/services/artifactregistry 6.274s
```

beta:
```
~/go/src/github.com/hashicorp/terraform-provider-google-beta$ make testacc TEST=./google-beta/services/artifactregistry TESTARGS='-run=TestAccDataSourceGoogleArtifactRegistryLocations_basic'
==> Checking that code complies with gofmt requirements...
go vet
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta/services/artifactregistry -v -run=TestAccDataSourceGoogleArtifactRegistryLocations_basic -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccDataSourceGoogleArtifactRegistryLocations_basic
=== PAUSE TestAccDataSourceGoogleArtifactRegistryLocations_basic
=== CONT  TestAccDataSourceGoogleArtifactRegistryLocations_basic
--- PASS: TestAccDataSourceGoogleArtifactRegistryLocations_basic (6.69s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta/services/artifactregistry       6.815s
```

**Comment by attwad** [comment]
ok! I have no idea what was happening but I deleted the provider forks and restarted from scratch, I could reproduce the error and finally got all lints/build/tests pass locally.

**Comment by attwad** [comment]
The VCR test that's failing seems to be unrelated to this PR (it's about the docker image data source, not the locations data source), is this good to go?

**Review by NickElliot — APPROVED** [review]
verified error is part of seperate issue https://github.com/hashicorp/terraform-provider-google/issues/18955

LGTM!

**Pull request merged**